### PR TITLE
Battery Message -> master

### PR DESCRIPTION
**Pull request by YHerm**

### CODE REVIEW

File: src/main/java/frc/utils/battery/BatteryLimiter.java [hunk 1]
```diff
@@ -3,18 +3,44 @@
...
 import frc.utils.DriverStationUtils;
+import frc.utils.dashboard.LoggedTableBoolean;
 import org.littletonrobotics.junction.Logger;
 
+import java.nio.file.Path;
+
 class BatteryLimiter extends Command {
 
     private static final int NUMBER_OF_SAMPLES_TAKEN_IN_AVERAGE = 50;
+    private static final double MESSAGE_STARTUP_TIME = 0.2; // 2 python cycles
```
Comment (tnrz135) [inline src/main/java/frc/utils/battery/BatteryLimiter.java:19]: I would prefer to split this to 2 constants of python cycle time and number of cycles to wait.

Also, what do you mean by python cycle time?

File: src/main/java/frc/utils/battery/BatteryLimiter.java [hunk 2]
```diff
@@ -34,10 +60,16 @@ public void execute() {
...
-            if (GlobalConstants.ENABLE_BATTERY_LIMITER && !DriverStationUtils.isMatch()) {
+            showBatteryMessage();
+            if (GlobalConstants.ENABLE_BATTERY_LIMITER && !DriverStationUtils.isMatch() && passedMessageStartUpTime()) {
                 throw new java.lang.RuntimeException("BATTERY IS LOW");
             }
         }
+        else {
+            if (isBatteryLow.get()) {
+                isBatteryLow.set(false);
+            }
```
Comment (tnrz135) [inline src/main/java/frc/utils/battery/BatteryLimiter.java:71]: CTRL Alt L

File: src/main/java/frc/utils/battery/BatteryLimiter.java
Comment (tnrz135) [inline src/main/java/frc/utils/battery/BatteryLimiter.java:1]: I get the idea with the image, but it's not good enough. While testing the robot during the season, you will look on the robot with a hand on the keyboard. You can very easily miss the warning or not be able to stop in time.

**Review by tnrz135 — CHANGES_REQUESTED** [review]

File: src/main/java/frc/utils/battery/BatteryLimiter.java (hunk 1 above)
Comment (YHerm) [reply src/main/java/frc/utils/battery/BatteryLimiter.java:19]: The loop in python runs every 100 ms

File: src/main/java/frc/utils/battery/BatteryLimiter.java
Comment (YHerm) [reply src/main/java/frc/utils/battery/BatteryLimiter.java:1]: Must of the times the code will throw too, but if not I believe you will notice pop up window, if not I don’t now what also we can do

File: src/main/java/frc/utils/battery/BatteryLimiter.java (hunk 1 above)
Comment (YHerm) [reply src/main/java/frc/utils/battery/BatteryLimiter.java:19]: And we check connection to robot and then check if we should show the message, so we need to wait 2 cycles for it to surely be able to show message before stopping

File: src/main/java/frc/utils/battery/BatteryLimiter.java [hunk 3]
```diff
@@ -34,10 +61,14 @@ public void execute() {
         double currentAverageVoltage = voltageFilter.calculate(BatteryUtils.getCurrentVoltage());
         if (currentAverageVoltage <= BatteryUtils.MIN_VOLTAGE) {
             reportLowBattery();
-            if (GlobalConstants.ENABLE_BATTERY_LIMITER && !DriverStationUtils.isMatch()) {
+            showBatteryMessage();
+            if (GlobalConstants.ENABLE_BATTERY_LIMITER && !DriverStationUtils.isMatch() && passedMessageStartUpTime()) {
```
Comment (bossyshrimp68) [inline src/main/java/frc/utils/battery/BatteryLimiter.java:65]: you check if its a match after you do the popup... you should still state that its low but not with a massive Noam

File: src/main/python/BatteryMessage.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
+import sys
+import time
+import tkinter as tk
```
Comment (YoniKiriaty) [inline src/main/python/BatteryMessage.py:3]: consider not using `as`

File: src/main/java/frc/utils/battery/BatteryLimiter.java [hunk 5]
```diff
@@ -34,9 +61,14 @@ public void execute() {
 		double currentAverageVoltage = voltageFilter.calculate(BatteryUtils.getCurrentVoltage());
 		if (currentAverageVoltage <= BatteryUtils.MIN_VOLTAGE) {
 			reportLowBattery();
-			if (GlobalConstants.ENABLE_BATTERY_LIMITER && !DriverStationUtils.isMatch()) {
-				throw new java.lang.RuntimeException("BATTERY IS LOW");
+			if (!DriverStationUtils.isMatch()) {
+				showBatteryMessage();
+				if (GlobalConstants.ENABLE_BATTERY_LIMITER && passedMessageStartUpTime()) {
+					throw new java.lang.RuntimeException("BATTERY IS LOW");
```
Comment (bossyshrimp68) [inline src/main/java/frc/utils/battery/BatteryLimiter.java:67]: import

**Review by bossyshrimp68 — CHANGES_REQUESTED** [review]

**Review by bossyshrimp68 — APPROVED** [review]

**Review by bossyshrimp68 — APPROVED** [review]

File: src/main/java/frc/utils/battery/BatteryLimiter.java
Comment (tnrz135) [reply src/main/java/frc/utils/battery/BatteryLimiter.java:1]: My issue is that the code will throw...

**Review by tnrz135 — CHANGES_REQUESTED** [review]

**Review by tnrz135 — DISMISSED** [review]

File: src/main/python/BatteryMessage.py [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+from tkinter import PhotoImage
+from NetworkTableManager import NetworkTableClient
+
+IMAGE_PATH = "../../../noam-battery-message.png"
+WINDOW_NAME = "Battery Message"
+
+CLIENT_NAME = "BatteryMessage"
+LOW_BATTERY_TOPIC_NAME = "BatteryMessage/LowBattery"
+
+TIME_BETWEEN_MESSAGES_SECONDS = 4
```
Comment (YHerm) [inline src/main/python/BatteryMessage.py:13]: Change to 60

Comment (YHerm) [reply src/main/python/BatteryMessage.py:13]: will be last change, needed for testing

**Pull request merged**